File: educative/grokking_coding_interview/in_place_reversal_of_ll/ReverseLinkedList.java
package grokking_coding_interview.in_place_reversal_of_ll;

class ReverseLinkedList {

    public static ListNode reverse(ListNode head) {
        ListNode current = head, previous = null, next = null;
        while (current != null) {
            next = current.next;
            current.next = previous;
            previous = current;
            current = next;
        }
        return previous;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(2);
        head.next = new ListNode(4);
        head.next.next = new ListNode(6);
        head.next.next.next = new ListNode(8);
        head.next.next.next.next = new ListNode(10);

        ListNode result = ReverseLinkedList.reverse(head);
        System.out.print("Nodes of the reversed LinkedList are: ");
        while (result != null) {
            System.out.print(result.value + " ");
            result = result.next;
        }
    }
}

class ListNode {
    int value = 0;
    ListNode next;

    ListNode(int value) {
        this.value = value;
    }
}

File: taro/reverse_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reverse_linked_list(head: ListNode) -> ListNode:
    previous = None
    current = head

    while current:
        next = current.next
        current.next = previous
        previous = current
        current = next
    return previous


def display(head: ListNode) -> None:
    current = head
    while current:
        if current.next:
            print(current.val, end= ' -> ')
        else:
            print(current.val, end='')

        current = current.next
    print()


if __name__ == '__main__':
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    display(head)
    display(reverse_linked_list(head))
    
    head = ListNode(1)
    head.next = ListNode(2)
    display(head)
    display(reverse_linked_list(head))

    head = ListNode(1)
    display(head)
    display(reverse_linked_list(head))
